### PR TITLE
Allow preinitializing spans created from arrays

### LINKED ISSUE #107541: [NativeAOT] Static constructor interpreter does not support spans
### Description

Given a program that statically initializes a lookup table while using a span to slice it during construction (extracted from https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/revcomp-csharpcore-8.html):
```cs
static class Program {
    static void Main() {
        // Simulate work
        var res = 0;
        for (var i = 0; i < LFSkipsAndReverseTable.Value.Length; i++)
            res += LFSkipsAndReverseTable.Value[i];
        Console.WriteLine(res);
    }
}

static class LFSkipsAndReverseTable {
    public static readonly byte[] Value;

    static LFSkipsAndReverseTable() {
        var vectors = new byte[(61 + 16) * 16];
        for (var i = 0; i < 16 + 61; i++) // linefeed skips
            vectors[i] = (byte)(i < 61 ? 16 : 15);
        for (var i = 16 * 16; i < 61 * 16; i++)
            vectors[i] = (byte)(15 - i % 16);
        var offset = vectors.AsSpan()[(61 * 16)..]; // <-- slicing the buffer
        for (var i = 0; i < 16; i++)
            for (var j = 0; j < 16; j++)
                offset[i * 16 + j] =
                    (byte)(j == 0 ? i : j < 16 - i ? 16 - j : 15 - j);
        Value = vectors;
    }
}
```
It appears that cctor interpreter bails on this implementation, resulting in lazy initialization with cctor check in `Main`:
```asm
G_M000_IG01:                ;; offset=0x0000
       push     rsi
       push     rbx
       sub      rsp, 40
G_M000_IG02:                ;; offset=0x0006
       xor      ebx, ebx
       xor      esi, esi
       lea      rcx, [(reloc 0x4201f8)]
       cmp      qword ptr [rcx-0x08], 0
       jne      SHORT G_M000_IG07 ;; <-- cctor check
G_M000_IG03:                ;; offset=0x0018
       mov      rcx, qword ptr [(reloc 0x420200)]
       mov      rax, gword ptr [rcx+0x08]
       cmp      dword ptr [rax+0x08], 0
       jle      SHORT G_M000_IG05
       align    [7 bytes for IG04]
G_M000_IG04:                ;; offset=0x0030
       mov      rax, gword ptr [rcx+0x08]
       cmp      esi, dword ptr [rax+0x08] ;; <-- length read
       jae      SHORT G_M000_IG08 ;; <-- bounds check
       movzx    rax, byte  ptr [rax+rsi+0x10]
       add      ebx, eax
       inc      esi
       mov      rax, gword ptr [rcx+0x08]
       cmp      dword ptr [rax+0x08], esi ;; <-- length read
       jg       SHORT G_M000_IG04
G_M000_IG05:                ;; offset=0x004B
       mov      ecx, ebx
       call     System.Console:WriteLine(int)
       nop      
G_M000_IG06:                ;; offset=0x0053
       add      rsp, 40
       pop      rbx
       pop      rsi
       ret      
G_M000_IG07:                ;; offset=0x005A
       call     CORINFO_HELP_READYTORUN_GCSTATIC_BASE
       jmp      SHORT G_M000_IG03
G_M000_IG08:                ;; offset=0x0061
       call     CORINFO_HELP_RNGCHKFAIL
       int3
```

However, if the cctor is modified in the following way:
```cs
static LFSkipsAndReverseTable() {
    var vectors = new byte[(61 + 16) * 16];
    for (var i = 0; i < 16 + 61; i++) // linefeed skips
        vectors[i] = (byte)(i < 61 ? 16 : 15);
    for (var i = 16 * 16; i < 61 * 16; i++)
        vectors[i] = (byte)(15 - i % 16);
    var offset = 61 * 16; // <-- replace slicing with offset
    for (var i = 0; i < 16; i++)
        for (var j = 0; j < 16; j++)
            vectors[offset + (i * 16 + j)] = // index into the original buffer with offset instead of using a span
                (byte)(j == 0 ? i : j < 16 - i ? 16 - j : 15 - j);
    Value = vectors;
}
```
Then `Main` compiles to the expected loop over pre-initialized array of constant length:
```asm
G_M000_IG01:                ;; offset=0x0000
       sub      rsp, 40
G_M000_IG02:                ;; offset=0x0004
       xor      ecx, ecx
       xor      eax, eax
       align    [0 bytes for IG03]
G_M000_IG03:                ;; offset=0x0008
       lea      rdx, gword ptr [(reloc 0x4201e8)]
       movzx    rdx, byte  ptr [rdx+rax+0x10]
       add      ecx, edx
       inc      eax
       cmp      eax, 0x4D0
       jl       SHORT G_M000_IG03
G_M000_IG04:                ;; offset=0x001F
       call     System.Console:WriteLine(int)
       nop      
G_M000_IG05:                ;; offset=0x0025
       add      rsp, 40
       ret
```

Ideally, the users should not have to know about this quirk in order to rely on pre-computation of data in such scenarios.  
It would be great to have cctor interpreter improved with the support of spans.  
Thanks!

### Configuration

```
.NET SDK:
 Version:           9.0.100-rc.2.24452.3
 Commit:            c294e97a49
 Workload version:  9.0.100-manifests.ba8980d7
 MSBuild version:   17.12.0-preview-24429-01+01e53f416
 ```

### Regression?
No

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Fixes #107541.

Cc @dotnet/ilc-contrib

### CODE REVIEW

**Comment by neon-sunset** [comment]
Thank you!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs [hunk 1]
```diff
@@ -612,6 +612,12 @@ private Status TryScanMethod(MethodIL methodIL, Value[] parameters, Stack<Method
 
                                 instance = new ReadOnlySpanValue(readOnlySpanElementType, bytes, byteOffset, byteLength);
                             }
+                            else if (readOnlySpanElementType != null && ctorSig.Length == 1 && ctorSig[0].IsArray
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs:615]: It took me some extra time to see that `readOnlySpanElementType` is set in the previous if condition. Would it be better to have all span ctors under the same if check? Something like:

```
else if ((TryGetSpanElementType(owningType, isReadOnlySpan: true, out MetadataType readOnlySpanElementType)
    || TryGetSpanElementType(owningType, isReadOnlySpan: false, out readOnlySpanElementType))
{
    if (ctorSig.Length == 1 ...)
    {
...
    }
    else if (ctorSig.Length == 2 ...)
    {
...
    }
}
```

**Review by jkotas — APPROVED** [review]

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs (hunk 1 above)
Comment (agocke) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs:615]:
```suggestion
                            else if (readOnlySpanElementType != null && ctorSig is [ { IsArray: true }]
```

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs:615]: Yeah, I don't like it either, but we'd need a `goto` to line 623 (in the new file, or 617 in the before version). If it's a newobj on readonly span that we don't intrinsically recognize, it needs to go to the "usual interpreter handling".

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs [hunk 2]
```diff
@@ -2929,6 +2935,18 @@ public void WriteContent(ref ObjectDataBuilder builder, ISymbolNode thisNode, No
                 builder.EmitBytes(_data);
             }
 
+            public bool TryGetReadOnlySpan(out ReadOnlySpanValue value)
+            {
+                var parameterType = ((ArrayType)Type).ParameterType as MetadataType;
+                if (parameterType != null)
```
Comment (agocke) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs:2941]:
```suggestion
                if (((ArrayType)Type).ParameterType is MetadataType parameterType)
```

**Review by agocke — APPROVED** [review]
A few suggestions on how you could use patterns to remove some repeated conditionals.

Feel free to ignore this feedback. Sometimes I find this useful, sometimes it reads worse. It's a judgement call.

**Comment by am11** [comment]
Not a regression but could it be a 9.0 candidate?

**Comment by MichalStrehovsky** [comment]
> Not a regression but could it be a 9.0 candidate?

We have a formal list of things that are allowed into 9.0 at this point and the only perf fixes that would meet 9.0 bar are "Significant performance regression from last release" (not just regression, it would also need to be significant to even consider a backport).

**Pull request merged**